src/Microsoft.AspNetCore.Authentication.OpenIdConnect/OpenIdConnectHandler.cs

```diff
@@ -520,6 +520,17 @@ protected override async Task<HandleRequestResult> HandleRemoteAuthenticateAsync
                 // if any of the error fields are set, throw error null
                 if (!string.IsNullOrEmpty(authorizationResponse.Error))
                 {
+                    // Note: access_denied errors are special protocol errors indicating the user didn't
+                    // approve the authorization demand requested by the remote authorization server.
+                    // Since it's a frequent scenario (that is not caused by incorrect configuration),
+                    // denied errors are handled differently using a special "access denied" exception.
+                    // Visit https://tools.ietf.org/html/rfc6749#section-4.1.2.1 for more information.
+                    if (string.Equals(authorizationResponse.Error, "access_denied", StringComparison.Ordinal))
+                    {
+                        return HandleRequestResult.Fail(new AccessDeniedException(
```

Comment:
I'd rather invoke the AccessDenied event here than create a new AccessDeniedException that's never actually thrown.

Reply:
That's possible, but it means we'll have the same (duplicated) logic in all the handlers. It will also be a little weird, as `RemoteAuthenticationEvents`/`RemoteAuthenticationOptions` will have events/options for something that is actually implemented in subclasses.

It's unfortunate, but errors are represented by `Exception`s in 2.0. I'd also have preferred a proper `AuthentiationError`, but I guess that's too late for that.

Reply:
Calling the event in a subclass is fine, OAuth CreatingTicket is a lot like that. You could share the code via a protected method if you like.